dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -2193,6 +2239,458 @@ const char* DynamicData::typekind_to_string(TypeKind tk) const
+        return false;
+      }
+    }
+    indent = temp_indent;
+    break;
+  }
+  case XTypes::TK_ARRAY: {
+    DCPS::String temp_indent = indent;
+    indent += "  ";
+    type_name = dd.type()->get_descriptor().element_type->get_descriptor().name;
+    member_name = dd.type()->get_descriptor().name;
+    XTypes::LBound bound = dd.type()->get_descriptor().bound[0];
+    type_string += "  " + type_name + "[" + DCPS::to_dds_string(bound) + "] " +  member_name + " =\n";
+    for (ACE_CDR::ULong i = 0; i < bound; ++i) {
+      type_string += indent + "[" + DCPS::to_dds_string(i) + "]";
+      if (dd.get_complex_value(temp_dd, i) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_complex_value\n"));
+        }
+        return false;
+      }
+      if (!print_dynamic_data(temp_dd, type_string, indent)) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to read struct member\n"));
+        }
+        return false;
+      }
+    }
+    indent = temp_indent;
+    break;
+  }
+  case XTypes::TK_STRUCTURE: {
+    DCPS::String temp_indent = indent;
+    indent += "  ";
+    type_string += "struct " + dd.type()->get_name() + " {\n";
+    XTypes::DynamicTypeMembersById dtmbi;
+    dd.type()->get_all_members(dtmbi);
+    for (XTypes::DynamicTypeMembersById::iterator iter = dtmbi.begin(); iter != dtmbi.end(); ++iter) {
+      dd.get_complex_value(temp_dd, iter->first);
+      member_name = iter->second->get_descriptor().name;
+      type_name = iter->second->get_descriptor().get_type()->get_descriptor().name;
+      type_string += indent + type_name + " " + member_name;
+      if (!print_dynamic_data(temp_dd, type_string, indent)) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to read struct member\n"));
+        }
+        return false;
+      }
+    }
+    indent = temp_indent;
+    type_string += indent + "};\n";
+    break;
+  }
+  case XTypes::TK_UNION: {
+    DCPS::String temp_indent = indent;
+    indent += "  ";
+    type_string += "union " + dd.type()->get_name() + " {\n";
+    ACE_CDR::ULong item_count = dd.get_item_count();
+    member_name = dd.type()->get_descriptor().discriminator_type->get_descriptor().name;
+    type_string += indent + member_name + " discriminator";
+
+    switch (dd.type()->get_descriptor().discriminator_type->get_kind()) {
+    case XTypes::TK_INT8: {
+      ACE_CDR::Int8 my_int8;
+      if (dd.get_int8_value(my_int8, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_int8_value\n"));
+        }
+        return false;
+      }
+      type_string += " = " + DCPS::to_dds_string(my_int8) + "\n";
+      break;
+    }
```

Comment:
Can we factor out the code in these cases to share with the code in the outer `switch`? Looks like they are similar enough.